[jit] Add removeAttribute to `ClassType`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28985 [quant][graphmode] Remove observer module after insert_quant_dequant
* #29048 [jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object
* **#28984 [jit] Add removeAttribute to `ClassType`**

Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

jerryzh168 committed:
[jit] Add removeAttribute to `ClassType`

Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -92,6 +92,22 @@ size_t ClassType::addAttribute(
+  std::vector<TypePtr> attributeTypes;
+  auto& names = attributeNames_;
+  auto& types = attributeTypes_;
+  TORCH_CHECK(std::find(names.begin(), names.end(), name) != names.end(), "Can't remove a non-existent attribute");
```
Please use getAttributeSlot, it does the assertion and the find.

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1476,6 +1476,11 @@ struct CAFFE2_API ClassType : public NamedType {
+  // [Internal Only] Remove attribute from the ClassType
+  // caller is responsible to make sure the attribute is
+  // no longer used
+  void removeAttribute(const std::string& name);
```
Let's put the warning in the function name: `unsafeRemoveAttribute`. Let's also add what it means for the attribute to be no longer used: it is unsafe  to having existing allocations of this object around anymore, and any code that works on the attribute is now invalid. Only newly created code is valid again.

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -92,6 +92,22 @@ size_t ClassType::addAttribute(
+  auto& names = attributeNames_;
+  auto& types = attributeTypes_;
+  TORCH_CHECK(std::find(names.begin(), names.end(), name) != names.end(), "Can't remove a non-existent attribute");
+  for (size_t i = 0; i < names.size(); ++i) {
```
These can just be calls to attributeNames_.erase, etc.

zdevito reviewed (commented):
See inline comments, also needs tests.

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

jerryzh168 commented:
@zdevito for test, I didn't find existing tests for `ClassType`, should I create a new test file?

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

suo commented:
> @zdevito for test, I didn't find existing tests for `ClassType`, should I create a new test file?

There is test_class_import and test_class_parser, you can make a new file that looks like those

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

jerryzh168 commented:
please take a look again @zdevito @suo

zdevito reviewed (approved):
Looks good, thanks!

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add removeAttribute to `ClassType`"


Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

[ghstack-poisoned]
